ai-models/tracker/sam2/func.py

```diff
@@ -0,0 +1,171 @@
+
+import collections
+import dataclasses
+
+import cvat_sdk.auto_annotation as cvataa
+import PIL.Image
+import torch
+import torchvision.transforms
+from cvat_sdk.masks import decode_mask, encode_mask
+from sam2.sam2_video_predictor import SAM2VideoPredictor
+from sam2.utils.misc import fill_holes_in_mask_scores
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class _PreprocessedImage:
+    original_width: int
+    original_height: int
+    vision_feats: list[torch.Tensor]
+    vision_pos_embeds: list[torch.Tensor]
+    feat_sizes: list[tuple[int, int]]
+
+
+@dataclasses.dataclass(kw_only=True)
+class _TrackingState:
+    frame_idx: int
+    predictor_outputs: dict
+
+
+class _Sam2Tracker:
+    def __init__(self, model_id: str, device: str = "cpu", **kwargs) -> None:
+        self._device = torch.device(device)
+        self._predictor = SAM2VideoPredictor.from_pretrained(
+            model_id, device=self._device, **kwargs
+        )
+        self._transform = torchvision.transforms.Compose(
+            [
+                torchvision.transforms.Resize(
+                    (self._predictor.image_size, self._predictor.image_size)
+                ),
+                torchvision.transforms.ToTensor(),
+                torchvision.transforms.Normalize(
+                    # see load_video_frames in the SAM2 source
+                    mean=(0.485, 0.456, 0.406),
+                    std=(0.229, 0.224, 0.225),
+                ),
+            ]
+        )
+
+    spec = cvataa.TrackingFunctionSpec(supported_shape_types=["mask"])
+
+    @torch.inference_mode()
+    def preprocess_image(
+        self, context: cvataa.TrackingFunctionContext, image: PIL.Image.Image
+    ) -> _PreprocessedImage:
+        image = image.convert("RGB")
+
+        image_tensor = self._transform(image).unsqueeze(0).to(device=self._device)
+        backbone_out = self._predictor.forward_image(image_tensor)
+        vision_feats = backbone_out["backbone_fpn"][-self._predictor.num_feature_levels :]
+        vision_pos_embeds = backbone_out["vision_pos_enc"][-self._predictor.num_feature_levels :]
+
+        return _PreprocessedImage(
+            original_width=image.width,
+            original_height=image.height,
+            vision_feats=[x.flatten(2).permute(2, 0, 1) for x in vision_feats],
+            vision_pos_embeds=[x.flatten(2).permute(2, 0, 1) for x in vision_pos_embeds],
+            feat_sizes=[(x.shape[-2], x.shape[-1]) for x in vision_pos_embeds],
+        )
+
+    def _call_predictor(self, *, pp_image: _PreprocessedImage, frame_idx: int, **kwargs) -> dict:
+        out = self._predictor.track_step(
+            current_vision_feats=pp_image.vision_feats,
+            current_vision_pos_embeds=pp_image.vision_pos_embeds,
+            feat_sizes=pp_image.feat_sizes,
+            point_inputs=None,
+            frame_idx=frame_idx,
+            num_frames=frame_idx + 1,
+            **kwargs,
+        )
+
+        return {
```

Comment:
Maybe I am lack of information here, but if you store state somewhere in redis, should not these tensors be converted GPU -> CPU?

Reply:
Currently the agent does not persist the states - it just stores them in memory and gives them back to the function as-is. When I implement persistence, I'm planning to add a dedicated method for serializing states, and it'll be responsible for moving the data to the CPU.